examples/local/vstream_client.go

```diff
@@ -82,17 +72,372 @@ func main() {
+			err = ctx.Err()
+			if err != nil {
+				return fmt.Errorf("context error: %w", err)
+			}
+
+			// keep track of the last event time for heartbeat monitoring. We're purposefully not using the event
+			// timestamp, since that would cause cancellation if the stream was copying, delayed, or lagging.
+			lastEventReceivedAtUnix.Store(time.Now().Unix())
+
+			// handle individual events based on their type
+			switch ev.Type {
+			// field events are sent first, and contain schema information for any tables that are being streamed,
+			// so we cache the fields for each table as they come in
+			case binlogdatapb.VEventType_FIELD:
+				switch ev.RowEvent.TableName {
+				case "customer":
+					customerFields = ev.FieldEvent.Fields
+
+				case "corder":
+					corderFields = ev.FieldEvent.Fields
+
+				default:
+					return errors.New("unexpected table name: " + ev.RowEvent.TableName)
+				}
+
+			// row events are the actual data changes, and we'll process them based on the table name
+			case binlogdatapb.VEventType_ROW:
+				// since our filter is "select * from customer", we could rely on that and not check the table name,
+				// but this shows how you might handle multiple tables in the same stream
+				switch ev.RowEvent.TableName {
+				case "customer":
+					var customer *Customer
+					customer, err = processCustomerRowEvent(customerFields, ev.RowEvent)
+					if err != nil {
+						return err
+					}
+
+					// we're not handling deletes, so we'll ignore nil customer return values
+					if customer != nil {
+						customers = append(customers, customer)
+					}
+
+				case "corder":
+					fmt.Printf("corder event: %v | fields: %v\n", ev.RowEvent, corderFields)
+					return errors.New("corder support not implemented")
+				}
+
+			// vgtid events are sent periodically, and we'll store the latest vgtid for checkpointing. We may get
+			// this mid-transaction, so we don't flush here, so we don't propagate a partial transaction that may
+			// be rolled back.
+			case binlogdatapb.VEventType_VGTID:
+				latestVgtid = ev.Vgtid
+
+			// commit and other events are safe to flush on, since they indicate the end of a transaction.
+			// Otherwise, there's not much to do with these events.
+			case binlogdatapb.VEventType_COMMIT, binlogdatapb.VEventType_OTHER:
+				// only flush when we have an event that guarantees we're not flushing mid-transaction
+				err = flushFunc()
+				if err != nil {
+					return err
+				}
+
+			// DDL events are schema changes, and we might want to handle them differently than data events.
+			// They are safe to flush on, since they indicate the end of a transaction. If you want to
+			// transparently adjust the destination schema based on DDL events, you would do that here.
+			case binlogdatapb.VEventType_DDL:
+				err = flushFunc()
+				if err != nil {
+					return err
+				}
+
+				// alter the destination schema based on the DDL event
+
+			case binlogdatapb.VEventType_COPY_COMPLETED:
```

Comment:
Should there be some sort of defer happening, that resets the destination tables and resets the last vgtid to empty?